hibernate-core/src/main/java/org/hibernate/metamodel/mapping/ordering/ast/DomainPathContinuation.java

```diff
@@ -64,9 +65,28 @@ public SequencePart resolvePathPart(
 					subPart
 			);
 		}
+		if ( referencedModelPart instanceof ToOneAttributeMapping ) {
+			ToOneAttributeMapping toOneAttribute = (ToOneAttributeMapping) referencedModelPart;
+			if ( name.equals( toOneAttribute.getTargetKeyPropertyName() ) ) {
+				final ModelPart subPart = toOneAttribute.findSubPart( name );
+				if ( subPart == null ) {
+					throw new PathResolutionException(
+							"Could not resolve path token : " + referencedModelPart + " -> " + name
+					);
+				}
+
+				return new DomainPathContinuation(
```

Comment:
Actually this is not totally correct and to be safe, I would suggest you introduce a new `SequencePart` implementation for `ToOneAttributeMapping` that is able to handle these special paths. The implementation would use the `targetKeyPropertyNames` and for every sub-part, prune possible further paths. If no paths are possible, this is not a simple FK reference and it should fail. The implementation could look roughly like this:

```
class FkDomainPathContinuation extends DomainPathContinuation {
	@Override
	public SequencePart resolvePathPart(
			String name,
			boolean isTerminal,
			TranslationContext translationContext) {
		List<String> newPossiblePaths = new ArrayList<>(possiblePaths.size());
		for (String path : possiblePaths) {
			if (path.startsWith(name)) {
				if (path.length() == name.length()) {
					return new DomainPathContinuation(
						navigablePath.append( name ),
						this,
						((ModelPartContainer) referencedModelPart).findSubPart(name, null)
					);
				}
				else if (path.charAt(name.length()) == '.') {
					newPossiblePaths.add(path.substring(name.length() + 1));
				}
			}
		}
		if (newPossiblePaths.isEmpty()) {
			throw new PathResolutionException(
				"Domain path of type `" +  referencedModelPart.getPartMappingType() +
				"Domain path of type `" + referencedModelPart.getPartMappingType() +
				"` -> `" + name + "`"
			);
		}
		return new FkDomainPathContinuation(
			navigablePath.append( name ),
			this,
			((ModelPartContainer) referencedModelPart).findSubPart(name, null),
			newPossiblePaths                                      
		);
	}
}
```

Reply:
Interesting. If I am not mistaken, we must also consider here the case in which the key is composed? I tried to address this comment with the commit https://github.com/hibernate/hibernate-orm/pull/4330/commits/d81e3c5c9114ec8dc4783da6d8aedea33b7832b6. I'm on the right track? Many thanks @beikov!